8293170: Improve encoding of the debuginfo nmethod section

bulasevich opened the pull request:
The nmethod "scopes data" section is 10% of the size of nmethod. Now the data is compressed using the Pack200 algorithm, which is good for encoding small integers (LineNumberTable, etc). Using the fact that half of the data in the partition contains zeros, I reduce its size by another 30%.
 
Testing: jtreg hotspot&jdk, Renaissance benchmarks

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293170](https://bugs.openjdk.org/browse/JDK-8293170): Improve encoding of the debuginfo nmethod section


### Reviewers
 * @Mkkebe (no known github.com user name / role) ⚠️ Review applies to [f365d780](https://git.openjdk.org/jdk/pull/10025/files/f365d78028eba5fc0a43841c56589e100e896102)
 * [Evgeny Astigeevich](https://openjdk.org/census#eastigeevich) (@eastig - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10025/head:pull/10025` \
`$ git checkout pull/10025`

Update a local copy of the PR: \
`$ git checkout pull/10025` \
`$ git pull https://git.openjdk.org/jdk pull/10025/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10025`

View PR using the GUI difftool: \
`$ git pr show -t 10025`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10025.diff">https://git.openjdk.org/jdk/pull/10025.diff</a>

</details>

dougxc commented:
I believe this change is JVMCI agnostic since it does not change the API of DebugInformationRecorder.

bulasevich commented:
> I believe this change is JVMCI agnostic since it does not change the API of DebugInformationRecorder.

Right. Thank you!

vnkozlov reviewed (commented):
Looks good. Let me test it.

dean-long commented:
What is the performance impact of making several of the methods virtual?

vnkozlov commented:
My builds failed:
```
src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedSparceDataReadStream.java:42: warning: [lossy-conversions] implicit cast from int to byte in compound assignment is possibly lossy
       b |= (0xFF & curr_byte_) >> (8 - byte_pos_);
                                ^
```

dougxc commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedSparceDataReadStream.java:
```diff
@@ -0,0 +1,73 @@
+
+import sun.jvm.hotspot.debugger.*;
+
+public class CompressedSparceDataReadStream extends CompressedReadStream {
```
CompressedSparceDataReadStream -> CompressedSparseDataReadStream

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,80 @@ void CompressedWriteStream::write_long(jlong value) {
+  return (jint)result;
+}
+
+int CompressedSparseDataWriteStream::position() {
```
The function with a side effect looks strange to me. I see an assert in `DebugInformationRecorder::DebugInformationRecorder(OopRecorder* oop_recorder)` which uses it for checking.  So the assert can cause side affects. I am not sure it is expected.

bulasevich replied:
Storing the debug info is an iterative process. Chunks of data are compared to avoid duplication, and on some points the generated data is discarded and position is unrolled. Besides read/write stream implementation internals, DebugInformationRecorder uses raw stream data access to track the similar chunks (see [DIR_Chunk](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/debugInfoRec.cpp#L57)) and [memcpy](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/nmethod.cpp#L1969) the raw data. We have either to (1) align the data on positions where the DebugInformationRecorder splits data into chunks or to (2) take the bit position into account.

I experimented with `int position` to contain both the stream bit position in the least significant bits and stream byte position in most significant bits. For me the code becomes less readable and the performance is questionable even without the DebugInformationRecorder update:
```
-  uint8_t b1 = _buffer[_position] << _bit_pos;
-  uint8_t b2 = _buffer[++_position] >> (8 - _bit_pos);
+  uint8_t b1 = _buffer[_position >> 3] << (_position & 0x7);
+  _position += 8;
+  uint8_t b2 = _buffer[_position >> 3] >> (8 - _position & 0x7);
```
I would avoid this change and stay with current implementation. In fact, there is not much aligned positions within the data. And `assert(_stream->position() > serialized_null, "sanity");` (thanks for noticing that!) in the constructor makes no problem because data is aligned at the beginning of the stream.

eastig replied:
If we introduce `flush()` we can explicitly call it before any accesses to the internal buffer: `buffer()` and `position()`. We will add asserts to `buffer()` and `position()` checking whether there is unflushed data. I always prefer being explicit.

bulasevich commented:
> What is the performance impact of making several of the methods virtual?

Good question! My experiments show that in the worst case, the performance of the debug write thread is reduced by 424->113 MB/s with virtual functions. Compared to compile time, this is miserable: сompilation takes 1000ms per method, while generation of 300 bytes of scopes data with virtual function (worst case) takes 3ms. And I do not see any regression with benchmarks.

bulasevich commented:
> My builds failed:
> 
> ```
> src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedSparceDataReadStream.java:42: warning: [lossy-conversions] implicit cast from int to byte in compound assignment is possibly lossy
>        b |= (0xFF & curr_byte_) >> (8 - byte_pos_);
>                                 ^
> ```

Yes. It was a collision with "8244681: Add a warning for possibly lossy conversion in compound assignments" change. I rebased my branch and fixed the issue. Now it should be Ok.

eastig commented:
@bulasevich, it would be useful to have some examples of cases either in the JBS issues or the PR.

eastig commented on src/hotspot/share/compiler/oopMap.hpp:
```diff
@@ -374,7 +374,7 @@ class OopMapStream : public StackObj {
   OopMapValue current()                 { return _omv; }
 #ifdef ASSERT
-  int stream_position() const           { return _stream.position(); }
+  int stream_position()                 { return _stream.position(); }
```
This change is an example that something is wrong with the design.
There is a concrete class `CompressedReadStream` with expected behaviour of `position`: no changes to `_stream`.  We have to break this contract to be able to compile `OopMapStream`.

bulasevich replied:
This change is unrolled as CompressedSparseDataReadStream is a separate class now (rework to avoid the virtual functions)

dean-long commented:
> > What is the performance impact of making several of the methods virtual?
> 
> Good question! My experiments show that in the worst case, the performance of the debug write thread is reduced by 424->113 MB/s with virtual functions. Compared to compile time, this is miserable: сompilation takes 1000ms per method, while generation of 300 bytes of scopes data with virtual function (worst case) takes 3ms. And I do not see any regression with benchmarks.

I was wondering more about read performance.  I would expect that the debuginfo could be read many more times than it is written.  Also, from 424 to 113 seems like a very large slowdown.

bulasevich commented:
> > > What is the performance impact of making several of the methods virtual?
> > 
> > 
> > Good question! My experiments show that in the worst case, the performance of the debug write thread is reduced by 424->113 MB/s with virtual functions. Compared to compile time, this is miserable: сompilation takes 1000ms per method, while generation of 300 bytes of scopes data with virtual function (worst case) takes 3ms. And I do not see any regression with benchmarks.
> 
> I was wondering more about read performance. I would expect that the debuginfo could be read many more times than it is written. Also, from 424 to 113 seems like a very large slowdown.

Right. With counters in virtual methods, I see that reading debug information is less frequent than writing. Anyway. Let me rewrite code without virtual functions.

bulasevich commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,92 @@ void CompressedWriteStream::write_long(jlong value) {
+
+void CompressedSparseDataWriteStream::grow() {
+  int nsize = _size * 2;
+  const int min_expansion = UNSIGNED5::MAX_LENGTH;
+  if (nsize < min_expansion*2) {
+    nsize = min_expansion*2;
+  }
```
I agree. Theoretically, one can call set_position() to something far outside the buffer capabilities, and get the write() to store data far beyond the buffer. As it does not happen really, let me add assert(nsize > _position, "sanity") here.

Also I remove (nsize < 2 * UNSIGNED5::MAX_LENGTH) check as the DebugInfo buffer [start size is 10K](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/debugInfoRec.cpp#L130).
By the way, 10K seems too big. It covers 999 cases out of 1000.

bulasevich commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,93 @@ class CompressedWriteStream : public CompressedStream {
+protected:
+  u_char* _buffer;
+  int     _position; // current byte offset
+  size_t  _byte_pos {0}; // current bit offset
```
Yes. I rename _byte_pos to _bit_pos

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,89 @@ void CompressedWriteStream::write_long(jlong value) {
+    write_zero();
+    return;
+  }
+  int bit0 = 0x80; // first byte upper bit is set to indicate a value is not zero
```
This is untraditional and confusing. In a byte we are preparing it is bit 7. The same comment is about `bit1`.

bulasevich replied:
Ok

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,89 @@ void CompressedWriteStream::write_long(jlong value) {
+    return;
+  }
+  int bit0 = 0x80; // first byte upper bit is set to indicate a value is not zero
+  juint next = val >> 6;
+  int bit1 = (next != 0) ? 0x40 : 0; // bit indicating a last byte
+  write_byte_impl(bit0 | bit1 | (val & 0x3f));
+  while (next != 0) {
+    bit1 = (next >> 7) ? 0x80 : 0;
+    write_byte_impl(bit1 | (next & 0x7f));
+    next >>= 7;
+  }
```
Could you please document the encoding/decoding schema in comments to the class? Could you also please include examples?
Why is it chosen to split 32 bits into 6, 7, 7, 7 and 5?

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,106 @@ void CompressedWriteStream::write_long(jlong value) {
+  // - the most frequent case (0 < x < 64) is encoded in one byte
+  // - the payload of the first byte is 6 bits, the payload of the following bytes is 7 bits
+  // - the most significant bit in the first byte is occupied by a zero flag
+  // - each byte has a bit indicating whether it is the last byte in the sequence
```
in each byte bit #6 indicates whether it is the last byte in the sequence

bulasevich replied:
such comment would be misleading

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,106 @@ void CompressedWriteStream::write_long(jlong value) {
+  if (_bit_pos == 0) {
+    return _position;
+  }
+  // flush current data and start a new byte
+  write(_curr_byte << (8 - _bit_pos));
+  _curr_byte = 0;
+  _bit_pos = 0;
```
Let's extract this and call it `flush()`.

bulasevich replied:
Ok. Now it is align() as _curr_byte is evicted

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,93 @@ class CompressedWriteStream : public CompressedStream {
+
+  int position(); // method have a side effect: the current byte becomes aligned
+  void set_position(int pos) {
+    position();
```
`flush()`

bulasevich replied:
OK

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,93 @@ class CompressedWriteStream : public CompressedStream {
+  int position(); // method have a side effect: the current byte becomes aligned
+  void set_position(int pos) {
+    position();
+    _position = pos;
```
I think `pos` must be `<= position()`. Should we check this?

bulasevich replied:
In fact, it is used as rollback. Though I do not want to limit functionality to this usage only.
Let me add `assert(_position < _size, "set_position is only used for rollback");` check here

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,106 @@ void CompressedWriteStream::write_long(jlong value) {
+  return _position;
+}
+
+void CompressedSparseDataWriteStream::write_zero() {
```
We can rewrite the function not to use `_curr_byte`. We work directly on `_buffer` and rename `_bit_pos` into `_used_bits`:
```
_buffer[_position] >>= 1;
if (++_used_bits == 8) {
  _position += 1;
  _used_bits = 0; 
}

bulasevich replied:
thanks for the idea not to use _curr_byte
_bit_pos is used in both Read and Write streams. I think proper name is _bit_position.

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,106 @@ void CompressedWriteStream::write_long(jlong value) {
+  }
+}
+
+void CompressedSparseDataWriteStream::write_byte_impl(uint8_t b) {
```
_buffer[_position] |= (b >> _used_bits);
_position += 1;
_buffer[_position] = (b << (8 - _used_bits));

bulasevich replied:
OK

eastig commented:
@bulasevich,
Could you please add gtest unit tests checking `CompressedSparseDataWriteStream`/`CompressedSparseDataReadStream`?

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,93 @@ class CompressedWriteStream : public CompressedStream {
   }
 };
 
+class CompressedBitStream : public ResourceObj {
```
Maybe it is better `CompressedSparseData`?

bulasevich replied:
OK

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,93 @@ class CompressedWriteStream : public CompressedStream {
+public:
+  CompressedSparseDataReadStream(u_char* buffer, int position) : CompressedBitStream(buffer, position) {}
+
+  void set_position(int pos) {
```
Are there uses of it? If no, let's remove it.

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,106 @@ void CompressedWriteStream::write_long(jlong value) {
 }
+
+
+bool CompressedSparseDataReadStream::read_zero() {
```
There is an issue here. If the last value written to a stream was 0, a reader would not know this is one 0 or eight 0s. Is there a guarantee that the number of reads will the same as the number of writes?

bulasevich replied:
Reader knows what he reads. We have no marks in a raw data.

eastig commented:
@bulasevich,
BTW, why do we have a lot of 0s? Is it a normal situation?

eastig commented on src/hotspot/share/code/debugInfo.hpp:
```diff
@@ -295,14 +295,14 @@ class MonitorValue: public ResourceObj {
 // debugging information. Used by ScopeDesc.
 
-class DebugInfoReadStream : public CompressedReadStream {
+class DebugInfoReadStream : public CompressedSparseDataReadStream {
```
I don't think `DebugInfoReadStream`/`DebugInfoWriteStream` need public inheritance. The relation  is more like composition.
I would have implemented them like:
```
class DebugInfoReadStream : private CompressedSparseDataReadStream {
public:
   // we are using only needed functions from CompressedSparseDataReadStream.
    using CompressedSparseDataReadStream::buffer();
    using CompressedSparseDataReadStream::read_int();
    using ...
};
```
Or
```
template <typename DataReadStream> class DebugInfoReadStream {
public:
    // define only needed functions which use a minimum number of functions from DataReadStream
};
```

I prefer the templates because we can easily switch between different implementations of `DataReadStream`/DataWriteStream` without doing this kind of modifications.

bulasevich replied:
@No templates please! :)
For me, the following change is counterproductive as well.
```
- class DebugInfoReadStream : private CompressedSparseDataReadStream {
+ class DebugInfoReadStream : private CompressedSparseDataReadStream {
+  public:
+   using CompressedSparseDataReadStream::read_int;
+   using CompressedSparseDataReadStream::read_signed_int;
+   using CompressedSparseDataReadStream::read_double;
+   using CompressedSparseDataReadStream::read_long;
+   using CompressedSparseDataReadStream::read_bool;
+   using CompressedSparseDataReadStream::CompressedSparseDataReadStream;
```

eastig commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedSparseDataReadStream.java:
```diff
@@ -0,0 +1,73 @@
+
+import sun.jvm.hotspot.debugger.*;
+
+public class CompressedSparseDataReadStream extends CompressedReadStream {
```
This needs to be aligned with C++ code.

bulasevich replied:
I have added a simple jtreg test. thanks!

eastig commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,121 @@ void CompressedWriteStream::write_long(jlong value) {
+}
+
+void CompressedSparseDataWriteStream::write_byte_impl(uint8_t b) {
+  if (_position >= _size) {
+    grow();
+  }
```
Now we have these checks spread across the code.
There are two actions changing `_postion`:
- `_position++`
- `set_position`

We can replace `_position++` with `inc_position` where we can have the check with `grow`.
Regarding `set_position `, I have looked at its current uses.
Its uses are to support shared debug info:
- We write info.
- We check if we have written the same info.
- If yes, we use the one written before and roll back position.

If I haven't missed other uses, the meaning of `set_position` is to roll back. In such  case, no `grow` is needed. I suggest to rename `set_position` to `roll_back_to` or `move_back_to`.

bulasevich replied:
(1) we have a few _position increments
(2) we have a few _buffer[_position] access places

The check `if (_position >= _size) { grow(); }` can go either to either after (1) of before (2). I prefer the latter because we do not need to extend a buffer if we do not write there.

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,98 @@ class CompressedWriteStream : public CompressedStream {
+    write_signed_int(high(value));
+  }
+
+  void flush() {
```
Why do we need `flush` if we modify data in place?

bulasevich replied:
now it is align()

bulasevich commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,106 @@ void CompressedWriteStream::write_long(jlong value) {
+void CompressedSparseDataWriteStream::write_byte_impl(uint8_t b) {
+  write((_curr_byte << (8 - _bit_pos)) | (b >> _bit_pos));
+  _curr_byte = (0xff >> (8 - _bit_pos)) & b;
+}
```
> I don't see what causes it to be written.
It was a flush side effect caused by get_position in the end of the buffer filling.
Anyway, it is rewritten now. Thanks!

bulasevich commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -209,6 +217,7 @@ void CompressedSparseDataWriteStream::write_int(juint val) {
 
 void CompressedSparseDataWriteStream::grow() {
   int nsize = _size * 2;
```
OK

bulasevich commented:
> @bulasevich, Could you please add gtest unit tests checking `CompressedSparseDataWriteStream`/`CompressedSparseDataReadStream`?
Yes. Thanks

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,102 @@ class CompressedWriteStream : public CompressedStream {
+    write_signed_int(high(value));
+  }
+
+  void align() {
+    if (_bit_position > 0) {
+      // start a new byte
+      _position++;
+      _bit_position = 0;
+    }
+  }
+  int position() {
+    align(); // method positon() have a side effect: the current data becomes aligned
+    return _position;
+  }
+  void set_position(int pos) {
```
I have an idea which might solve the issues.

bulasevich replied:
> // Roll the stream state back to the marked one.  
> void roll_back();

get_position() is not about roll back only. See the DebugInformationRecorder, it serializes offsets into the stream.

```
int DebugInformationRecorder::serialize_scope_values(...) {
  ...
  int result = stream()->position();
  ...
  return result;
}

DebugToken* DebugInformationRecorder::create_scope_values(...) {
  ...
  return (DebugToken*) (intptr_t) serialize_scope_values(values);
}

void PhaseOutput::Process_OopMap_Node(MachNode *mach, int current_offset) {
  ...
  DebugToken *locvals = C->debug_info()->create_scope_values(locarray);
  DebugToken *expvals = C->debug_info()->create_scope_values(exparray);
  DebugToken *monvals = C->debug_info()->create_monitor_values(monarray);

  C->debug_info()->describe_scope(
    ...
    locvals,
    expvals,
    monvals
  );
  
void DebugInformationRecorder::describe_scope(...
                                              DebugToken* locals,
                                              DebugToken* expressions,
                                              DebugToken* monitors) {
  ...
  // serialize the locals/expressions/monitors
  stream()->write_int((intptr_t) locals);
  stream()->write_int((intptr_t) expressions);
  stream()->write_int((intptr_t) monitors);
```

eastig replied:
Thank you for the information. It's very helpful.

I think we should not simulate `CompressedWriteStream`. 

`DebugInformationRecorder` needs certain operations:
We write debug info into a stream writer: as grouped multiple data and single data. We need to know where bytes of grouped data begin and end. We need to keep offsets of grouped data in the stream. We need to be able to discard last written grouped data. We need to get the number of used bytes. We don't need to know how data stored in a stream.

Based on the specification, we need a stream writer to provide operations:
```
// Start grouped data.
// Return a position (byte offset) in the stream where grouped data begins.
int start_group();

// Finish grouped data.
// Return a position (byte offset) in the stream where grouped data ends.
int finish_group();

// Revert the stream to the specified position.
void set_position(int pos);

// Return the number of bytes stored data uses.
int data_size() const;
```

With them we don't have a function which in one implementation is const but in another implementation is with side effects. IMHO, at some point later side effects will cause bugs. 

Possible implementations:
```
int start_group() {
  complete_current_byte(); // this is renamed align()
  return _position;
}

int finish_group() {
  complete_current_byte();
  return _position;
}

int data_size() const {
  if (_position == 0 && _bit_position == 0)
    return 0;

 int used_bytes = _position;
 if (_bit_position != 0)
   ++used_bytes;
 return used_bytes;
}
```

In `DebugInformationRecorder` we will need to replace `position()` with `start_group()` and add `finish_group()`.
We will need to change `int DebugInformationRecorder::find_sharable_decode_offset(int stream_offset)` to
`int DebugInformationRecorder::find_sharable_decode_offset(int data_begin_offset, int data_end_offset)`.

If we want `DebugInformationRecorder` to use `CompressedWriteStream` we can use an adapter:
```
class CompressedWriteStreamAdapter: public CompressedWriteStream {
public:
...
int start_group() {
  return position();
}

int finish_group() {
  return position();
}

int data_size() const {
  return position();
}
 
};
```

bulasevich replied:
Sorry for not being online for a while :)
As you proposed, I refactored the CompressedSparseDataWriteStream interface.
I removed position() and set_position(int) methods. I added start_scope() and roll_back(int) instead.
Plus I have added is_empty() method (without side effect) to avoid unintentional align from assert.
Plus I added data_size() method (without side effect) - it is used to check the amount of data to copy.
Please check if it is Ok now.
Thanks.

bulasevich commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,102 @@ class CompressedWriteStream : public CompressedStream {
+    _position = position;
+  }
+
+  u_char* buffer() const { return _buffer; }
```
Please check my change.
- I move _buffer from CompressedSparseData to CompressedSparseDataReadStream and CompressedSparseDataWriteStream - the first one is constant.
- I update debugInfoRec.cpp, nmethod.cpp, nmethod.hpp. On the one hand, I would like to avoid a lot of file chaining. On the other hand, the const modifier adds readability.
Is that what you meant?

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,122 @@ class CompressedWriteStream : public CompressedStream {
+  size_t  _bit_position {0}; // current bit offset
+
+public:
+  CompressedSparseData(int position = 0) {
```
As it is one-argument constructor, let it be `explicit CompressedSparseData`.

bulasevich replied:
ok

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,122 @@ class CompressedWriteStream : public CompressedStream {
+public:
+  // Start grouped data. Return a byte offset position in the stream where grouped data begins
+  int start_scope() {
+    align(); // a side effect!
```
I think we don't need the comment here.

bulasevich replied:
sure

eastig commented on src/hotspot/share/code/compressedStream.hpp:
```diff
@@ -111,4 +112,122 @@ class CompressedWriteStream : public CompressedStream {
+  void roll_back(int pos) {
+    _position = pos;
+    _bit_position = 0;
+    assert(_position < _size, "set_position is only used for rollback");
```
Should we change the assert and move it to the beginning of the function?
```
assert(pos <= _position, "new position must be rollback the current position"
```

bulasevich replied:
right. thank you

eastig reviewed (changes requested):
Hotspot/share/code looks good to me. Just a few minor changes.

bulasevich commented on src/hotspot/share/code/compressedStream.cpp:
```diff
@@ -112,3 +112,111 @@ void CompressedWriteStream::write_long(jlong value) {
+
+bool CompressedSparseDataReadStream::read_zero() {
+  if (_buffer[_position] & (1 << (7 - _bit_position))) {
+    return 0; // not a zero data
```
ok

bulasevich commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedSparseDataReadStream.java:
```diff
@@ -0,0 +1,75 @@
+    }
+    byte b = readByteImpl();
+    int result = b & 0x3f;
+    for (int i = 0; (0 != ((i == 0) ? (b & 0x40) : (b & 0x80))); i++) {
```
If you don't mind, I would prefer shorter version.
```
for (int i = 0; (0 != ((i == 0) ? (b & 0x40) : (b & 0x80))); i++) {
  b = readByteImpl();
  result |= ((b & 0x7f) << (6 + 7 * i));
}

->

int result = b & 0x3f;
if ((b & 0x40) != 0) {
    b = readByteImpl();
    result |= (b & 0x7f) << 6;
    for (int i = 1; (b & 0x80) != 0; ++i) {
          b = readByteImpl();
          result |= ((b & 0x7f) << (6 + 7 * i));
    }
}
```

bulasevich commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/code/CompressedSparseDataReadStream.java:
```diff
@@ -0,0 +1,75 @@
+    return (short)buffer.getCIntegerAt(position, 1, true);
+  }
+
+  public byte readByteImpl() {
```
ok

bulasevich commented on test/hotspot/jtreg/serviceability/sa/TestCompressedSparseDataReadStream.java:
```diff
@@ -0,0 +1,107 @@
+        assertEquals(in.readInt(), 0);  // zero bit -> 0
+        assertEquals(in.readInt(), 0);  // zero bit -> 0
+        in.setPosition(2);
+        assertEquals(in.readInt(), 48); // 0xf000   -> 48
```
ok

bulasevich commented:
@dean-long @vnkozlov 
I eliminated virtual methods, changed the implementation internals and added tests.
Let me once again ask you to review. Thanks

rose00 reviewed (changes requested):
I am sorry, but we need to pump the brakes here.  This proposal is not acceptable in its current form.

There's a classic tradeoff between writing specialized encodings for each kind of data, versus using shared encodings.  The specialized encodings might be more compact and/or faster to read or write because they are tuned for a particular kind of data.  The shared encodings might be less compact, etc., but because they are shared the work of learning, documenting, analyzing, and testing them is done once.  That's why people use LEB128 as a varint in so many places:  It is not the best varint for most purposes, but it is easy to learn and well-understood.  You know what you are getting when you use it.

These varint formats are rather easy to create.  Here we have one that has good statistics for a particular kind of data, in the cases available today we have tested.  But if we accept it into our source base, maintainers will have to learn to deal with it in the future.  If the statistics of the data change (this will surely happen at some point) then what is a nice local optimization will become an analysis problem, a black box that maintainers will not want to touch.

To address such maintenance problems, we have picked off-the-shelf algorithms, documented and implemented them carefully, and reused them in our code base.

In HotSpot, we are using UNSIGNED5 (from Pack200 in the Java ecosystem) as an internal varint, yet another one.  It is slightly better than LEB128.  (We also use UTF8, of course, for similar purposes.)  We have factored out the algorithm so that it can be shared in more places in HotSpot.  For example, there is a line of work where field-info structures will be compressed with var-ints.  And there may be more to come; the reloc-info streams come to my mind as candidates for re-encoding.

Having a centralized varint mechanism is the smart call, because it can be maintained on behalf of multiple use cases.  And if it needs improvement or tuning, such changes will benefit all users.

By contrast, using a local compression scheme for a particular kind of data, as proposed here, creates a separate account of technical debt in each place.

So, let's not.

Instead, if we notice that we have zero-rich data, let's first try to do something that composes with with the existing varint scheme (UNSIGNED5) and works on top of it, or underneath it.  That is, stack a zero-reduction scheme with our varint scheme, instead of simultaneously inventing a zero-reduction scheme and a new varint scheme.

As a very simple example of this, the first N (N=16, say) code points `J=[0..N-1]` of UNSIGNED5 can be special-cased to mean "there are J zero values here".  A byte J in that range means that there are J+1 zeroes to decode.  A byte beyond that range is first decoded as X0 and then adjusted as `X=X0-N+1`, so that it is a full-range 32-bit integer.  The last N-1 values in the range, `[1-N..-1]`, will decode from 5-byte encodings, using 32-bit overflow, which the UNSIGNED5 algorithm is tolerant of.

As another simple example, probably overkill but more similar to the present proposal, the first `N=2^K` code points could be reserved as above (say, `K=4,N=16`), and the decoder state would have not just a count of extra zero values, but rather a bitmask of them.  In that case, the bitmask for some byte `J<2^K` would always be initialized to `2*J+1` so that there is always at least one bit set.

And as a different suggestion, the most attractive off-the-shelf zero-reduction scheme I am aware of is in Capn Proto [1].

[1]: <https://capnproto.org/encoding.html#packing>

It is similar to the zero-reduction scheme proposed here, but is more battle-tested and probably better thought through.  I suggest we investigate its use, as a standard compression technique, to compress these streams, preferably as-is, or perhaps adapted; preferably in conjunction with our existing varint scheme, or stand-alone.

I suspect (though I am not certain) that Capn Proto's packing scheme would work fine as a byte source for UNSIGNED5.  It is organized as a word-wise algorithm, which allows it to use fully-loaded registers as temps.  Stacking the two together would involve streaming through packed words, sticking those words into a small buffer (one cache line is enough), and decoding from there as UNSIGNED5.  Again, doing something like this (or either of the two suggestions above), would compose known-good techniques, would reduce maintenance burden, would solve the problem at hand (probably), and would avoid the technical debt of having multiple overlapping varint schemes in our code base.

rose00 reviewed (commented):
P.S. One reason I know about the Capn Proto packing is as a candidate for fast streaming (de)compression of heap snapshots.  We don't have that feature today, but may in the future for CDS and/or Leyden, and all of my arguments about using off-the-shelf techniques will apply there as well.
